File: src/preprocessing/preprocess_richheader.scala
import com.datastax.spark.connector._
import java.io.{ByteArrayOutputStream, ByteArrayInputStream}
import java.util.zip.{GZIPOutputStream, GZIPInputStream}

case class RichheaderServiceName(service_name: String, sha256: String)
case class RichheaderLabel(sha256: String, source_tags: String)
case class RichheaderServiceNameAndLabel(sha256: String, service_name: String, source_tags: String)
case class RichheaderSha256(sha256: String, service_name: String, results: Array[Byte])
case class RichheaderJoin(sha256: String, service_name: String, results: String, source_tags: String)

val richheader_service_name_rdd = sc.cassandraTable[RichheaderServiceName](KEYSPACE, SERVICE_NAME_TABLE)
                                .where("service_name=?", "richheader")
                                .keyBy(x => (x.sha256));
val richheader_label_rdd = sc.cassandraTable[RichheaderLabel](KEYSPACE, OBJECTS_TABLE)
                         .keyBy(x => (x.sha256));
val richheader_service_name_and_label_rdd = richheader_service_name_rdd.join(richheader_label_rdd)
                                          .map(x =>(new RichheaderServiceNameAndLabel(x._1, x._2._1.service_name, x._2._2.source_tags)))
                                          .keyBy(x => (x.sha256, x.service_name));

val richheader_sha256_rdd = sc.cassandraTable[RichheaderSha256](KEYSPACE, SHA256_TABLE)
                          .where("service_name=?", "richheader")
                          .keyBy(x => (x.sha256, x.service_name));
val richheader_join = richheader_service_name_and_label_rdd.join(richheader_sha256_rdd)
                    .map(x => (new RichheaderJoin(x._1._1, x._1._2, unzip(x._2._2.results), x._2._1.source_tags)))
                    .distinct();

richheader_join.toDF().write.format("parquet").save("richheader.df");

File: src/preprocessing/preprocess_richheader.py
import json
import time

richheader_rdd = sqlContext.read.parquet(DF_LOCATION).rdd
richheader_results = richheader_rdd.map(lambda x: (x.sha256, x.service_name, find_compid_in_richheader(x.results), convert_to_labels(x.source_tags), int(time.time())) if 'benign' not in x.source_tags else None).filter(bool)
richheader_objects = richheader_rdd.map(lambda x: (x.sha256, find_compid_in_richheader(x.results), convert_to_labels(x.source_tags), int(time.time())) if 'benign' not in x.source_tags else None).filter(bool)

def find_compid_in_richheader(results):
    compid_list = []

    richheader = json.loads(results)

    if 'cmpids' not in richheader:
        return [0] * 80

    for compid in richheader['cmpids']:
        compid_list.extend([compid['mcv'], compid['pid'], compid['cnt']])

    compid_list += [0] * (80 - len(compid_list))

    return compid_list

def convert_to_labels(source_tags):
    labels = source_tags[1:len(source_tags) - 1].split(',')
    labels.remove('malicious')
    return labels

richheader_df = richheader_results.toDF()
richheader_df = richheader_df.withColumnRenamed("_1", "sha256").withColumnRenamed("_2", "service_name").withColumnRenamed("_3", "features").withColumnRenamed("_4", "labels").withColumnRenamed("_5", "timestamp")
richheader_df.write.format("org.apache.spark.sql.cassandra").mode('append').options(table=PREPROCESSING_RESULTS, keyspace=KEYSPACE).save()

richheader_df = richheader_objects.toDF()
richheader_df = richheader_df.withColumnRenamed("_1", "sha256").withColumnRenamed("_2", "features_richheader").withColumnRenamed("_3", "labels").withColumnRenamed("_4", "timestamp")
richheader_df.write.format("org.apache.spark.sql.cassandra").mode('append').options(table=PREPROCESSING_OBJECTS, keyspace=KEYSPACE).save()

